quic/s2n-quic-transport/src/space/datagram.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    endpoint,
+    stream::{AbstractStreamManager, StreamTrait as Stream},
+    transmission::WriteContext,
+};
+use s2n_quic_core::{
+    datagram::{Endpoint, Sender},
+    frame,
+};
+
+pub struct Manager<Config: endpoint::Config> {
+    sender: <<Config as endpoint::Config>::DatagramEndpoint as Endpoint>::Sender,
+    // TODO: Remove this warning once Receiver is implemented
+    #[allow(dead_code)]
+    receiver: <<Config as endpoint::Config>::DatagramEndpoint as Endpoint>::Receiver,
+}
+
+impl<Config: endpoint::Config> Manager<Config> {
+    pub fn new(
+        sender: <<Config as endpoint::Config>::DatagramEndpoint as Endpoint>::Sender,
+        receiver: <<Config as endpoint::Config>::DatagramEndpoint as Endpoint>::Receiver,
+    ) -> Self {
+        Self { sender, receiver }
+    }
+
+    // This function creates a packet that a user can interrogate to learn
+    // more about the packet space available for datagrams, and to write
+    // datagrams to the packet.
+    pub fn on_transmit<S: Stream, W: WriteContext>(
+        &mut self,
+        context: &mut W,
+        stream_manager: &mut AbstractStreamManager<S>,
+    ) {
+        let mut packet = Packet {
+            context,
+            pending_streams: stream_manager.pending_streams(),
+        };
+        self.sender.on_transmit(&mut packet);
+    }
+}
+struct Packet<'a, C: WriteContext> {
+    context: &'a mut C,
+    pending_streams: bool,
+}
+
+const FRAME_TYPE_LEN: usize = 1;
+
+impl<'a, C: WriteContext> s2n_quic_core::datagram::Packet for Packet<'a, C> {
+    /// Returns the remaining space in the packet
+    fn remaining_capacity(&self) -> usize {
+        self.context.remaining_capacity()
+    }
+
+    /// Returns the largest datagram that can fit in space remaining in the packet
+    fn maximum_datagram_payload(&self) -> usize {
```

Comment:
what's the need for both of these? It seems like `maximum_datagram_payload` is always `remaining_capacity` - 1?

Reply:
I know. I don't think both are necessary. Technically they do serve two different purposes, but it is a little ridiculous to have two. What I am trying to do is give the user accurate information, because any space that's in the buffer will have to account for the frame size. The user can probably infer the maximum datagram they could send based on the remaining_capacity function.

I think I should just delete one of the functions. It can be added in later, but right now it is a little ridiculous.